File: src/server/utils/currency.js
/**
 * Currency conversion utilities.
 * FX rates are always fetched server-side. Never call these from client components.
 */

const CACHE_TTL_MS = 60 * 60 * 1000; // 1 hour

const rateCache = new Map();

/**
 * @param {string} base - Base currency code (e.g. "USD")
 * @returns {Promise<Record<string, number>>}
 */
async function fetchRates(base) {
  const cacheKey = base.toUpperCase();
  const cached = rateCache.get(cacheKey);

  if (cached && Date.now() - cached.timestamp < CACHE_TTL_MS) {
    return cached.rates;
  }

  const apiKey = process.env.OPEN_EXCHANGE_RATES_KEY;
  if (!apiKey) throw new Error("OPEN_EXCHANGE_RATES_KEY is not set");

  const res = await fetch(
    `https://openexchangerates.org/api/latest.json?app_id=${apiKey}&base=${cacheKey}`
  );

  if (!res.ok) throw new Error(`FX fetch failed: ${res.status}`);

  const data = await res.json();
  rateCache.set(cacheKey, { rates: data.rates, timestamp: Date.now() });
  return data.rates;
}

/**
 * Convert an amount from one currency to another.
 * Returns null if either currency is unknown.
 * @param {number} amount
 * @param {string} from
 * @param {string} to
 * @param {Record<string, number>} rates - Rates relative to a common base
 * @returns {number|null}
 */
function convert(amount, from, to, rates) {
  if (amount === null || amount === undefined) return null;
  if (from === to) return amount;

  const fromRate = rates[from.toUpperCase()];
  const toRate = rates[to.toUpperCase()];

  if (!fromRate || !toRate) return null;

  return (amount / fromRate) * toRate;
}

/**
 * Format a monetary value with its currency code, always shown.
 * e.g. formatMoney(4800, "USD") → "$4,800 USD"
 * @param {number} amount
 * @param {string} currency
 * @returns {string}
 */
function formatMoney(amount, currency) {
  if (amount === null || amount === undefined) return `— ${currency}`;
  const formatted = new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: currency.toUpperCase(),
    maximumFractionDigits: 0,
  }).format(amount);
  return `${formatted} ${currency.toUpperCase()}`;
}

module.exports = { fetchRates, convert, formatMoney };

File: backend/utils/currency.py
"""
Currency conversion utilities.
FX rates always fetched server-side. Never expose the API key to the client.
Direct Python port of server/utils/currency.js.
"""

import os
import time
import requests

CACHE_TTL = 3600  # 1 hour

_rate_cache: dict[str, dict] = {}
_cache_timestamps: dict[str, float] = {}


def fetch_rates(base: str) -> dict:
    key = base.upper()
    now = time.time()

    if key in _rate_cache and now - _cache_timestamps.get(key, 0) < CACHE_TTL:
        return _rate_cache[key]

    api_key = os.environ.get("OPEN_EXCHANGE_RATES_KEY")
    if not api_key:
        raise RuntimeError("OPEN_EXCHANGE_RATES_KEY is not set")

    resp = requests.get(
        f"https://openexchangerates.org/api/latest.json",
        params={"app_id": api_key, "base": key},
        timeout=10,
    )
    resp.raise_for_status()
    data = resp.json()
    rates = data["rates"]

    _rate_cache[key] = rates
    _cache_timestamps[key] = now
    return rates


def convert(amount: float | None, from_: str, to: str, rates: dict) -> float | None:
    if amount is None:
        return None
    if from_.upper() == to.upper():
        return amount

    from_rate = rates.get(from_.upper())
    to_rate = rates.get(to.upper())

    if from_rate is None or to_rate is None:
        return None

    return (amount / from_rate) * to_rate


def format_money(amount: float | None, currency: str) -> str:
    if amount is None:
        return f"— {currency.upper()}"
    # Basic formatting — frontend handles locale-specific display
    return f"{amount:,.0f} {currency.upper()}"
